REGRESSION (275831@main): Music video can't be played after tapping music video in Music app

jyavenard opened the pull request:
#### d03d3a51627299d0e4de746b80889edf43fe93a5
<pre>
REGRESSION (275831@main): Music video can&apos;t be played after tapping music video in Music app
<a href="https://bugs.webkit.org/show_bug.cgi?id=275391">https://bugs.webkit.org/show_bug.cgi?id=275391</a>
<a href="https://rdar.apple.com/128055030">rdar://128055030</a>

Reviewed by Jer Noble.

Commit 275831@main exposed two long-standing and existing issue:
1- When the video is paused following a system interruption, currentTime can continue to progress
despite the new video rate being 0.
2- The HTMLMediaElement can incorrectly resume playback of a paused video.

1) was due to MediaPlayerPrivateAVFObjC::paused() returning false as it only checks if the AVPlayer
status was AVPlayerTimeControlStatusPaused. This caused the default `timeIsProgressing()`
implementation to return true ; hence the time estimation in the content process to continue.

2) When the PlatformMediaSession was receiving an interruption notification, it would be ignored
on start, but not on end, which would resume playback.

Add test.

* LayoutTests/media/picture-in-picture/video-playback-system-interruption-and-resume-expected.txt: Added.
* LayoutTests/media/picture-in-picture/video-playback-system-interruption-and-resume.html: Added.
* Source/WebCore/platform/audio/PlatformMediaSession.cpp:
(WebCore::PlatformMediaSession::activeInterruptionCount const): We add a flag indicating if the
notification was ignored, in order to determine if playback needs to be resumed later.
(WebCore::PlatformMediaSession::interruptionType const):
(WebCore::PlatformMediaSession::beginInterruption):
(WebCore::PlatformMediaSession::endInterruption):
(WebCore::PlatformMediaSession::blockedBySystemInterruption const):
* Source/WebCore/platform/audio/PlatformMediaSession.h: Add method.
(WebCore::PlatformMediaSession::interruptionCount const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::rateChanged): Add logging

Canonical link: <a href="https://commits.webkit.org/279977@main">https://commits.webkit.org/279977@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4ec4fc47384dad968a575a451bcd25d2ee8f5d9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55082 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7698 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58364 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5814 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57382 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42109 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5853 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3973 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57177 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32621 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47724 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25722 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29399 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5071 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3958 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59954 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30363 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5455 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52026 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31485 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47792 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51480 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12112 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32512 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31275 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/audio/PlatformMediaSession.h:
```diff
@@ -261,14 +261,17 @@ class PlatformMediaSession
-    int m_interruptionCount { 0 };
+    struct Interruption {
+        InterruptionType type;
+        bool ignored;
```
Nit: can these have default initializers?

jernoble commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
```diff
@@ -87,8 +87,8 @@ class MediaPlayerPrivateAVFoundationObjC final : public MediaPlayerPrivateAVFoun
-    void didEnd() final;
-    void metadataLoaded() final;
+    void didEnd() override;
+    void metadataLoaded() override;
```
Are all of these `final` -> `override` changes really necessary?

jernoble commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -533,6 +533,7 @@ bool MediaPlayerPrivateRemote::seeking() const
 
 void MediaPlayerPrivateRemote::rateChanged(double rate, MediaTimeUpdateData&& timeData)
 {
+    INFO_LOG(LOGIDENTIFIER, "rate:", rate, " currentTime:", timeData.currentTime, " timeIsProgressing:", timeData.timeIsProgressing);
```
Ditto.

jyavenard replied:
Not sure what this refers to

jernoble reviewed (approved):
r=me with nits.

jyavenard committed:
REGRESSION (275831@main): Music video can't be played after tapping music video in Music app
https://bugs.webkit.org/show_bug.cgi?id=275391
rdar://128055030

Reviewed by Jer Noble.

Commit 275831@main exposed two long-standing and existing issue:
1- When the video is paused following a system interruption, currentTime can continue to progress
despite the new video rate being 0.
2- The HTMLMediaElement can incorrectly resume playback of a paused video.

1) was due to MediaPlayerPrivateAVFObjC::paused() returning false as it only checks if the AVPlayer
status was AVPlayerTimeControlStatusPaused. This caused the default `timeIsProgressing()`
implementation to return true ; hence the time estimation in the content process to continue.

2) When the PlatformMediaSession was receiving an interruption notification, it would be ignored
on start, but not on end, which would resume playback.

Add test.

* LayoutTests/media/picture-in-picture/video-playback-system-interruption-and-resume-expected.txt: Added.
* LayoutTests/media/picture-in-picture/video-playback-system-interruption-and-resume.html: Added.
* Source/WebCore/platform/audio/PlatformMediaSession.cpp:
(WebCore::PlatformMediaSession::activeInterruptionCount const): We add a flag indicating if the
notification was ignored, in order to determine if playback needs to be resumed later.
(WebCore::PlatformMediaSession::interruptionType const):
(WebCore::PlatformMediaSession::beginInterruption):
(WebCore::PlatformMediaSession::endInterruption):
(WebCore::PlatformMediaSession::blockedBySystemInterruption const):
* Source/WebCore/platform/audio/PlatformMediaSession.h: Add method.
(WebCore::PlatformMediaSession::interruptionCount const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::rateChanged): Add logging

Canonical link: https://commits.webkit.org/279977@main
